tree.go

```diff
@@ -449,75 +451,39 @@ func (tree *Tree) newnode() (p *node) {
 	return &(tree.alloc[ln])
 }
 
-func loadip4(ipstr []byte) (uint32, error) {
-	var (
-		ip  uint32
-		oct uint32
-		b   byte
-		num byte
-	)
-
-	for _, b = range ipstr {
-		switch {
-		case b == '.':
-			num++
-			if 0xffffffff-ip < oct {
-				return 0, ErrBadIP
-			}
-			ip = ip<<8 + oct
-			oct = 0
-		case b >= '0' && b <= '9':
-			oct = oct*10 + uint32(b-'0')
-			if oct > 255 {
-				return 0, ErrBadIP
-			}
-		default:
-			return 0, ErrBadIP
-		}
-	}
-	if num != 3 {
-		return 0, ErrBadIP
-	}
-	if 0xffffffff-ip < oct {
-		return 0, ErrBadIP
-	}
-	return ip<<8 + oct, nil
-}
+func parseCIDR(cidr []byte) (net.IP, net.IPMask, error) {
+	var address netip.Addr
+	var prefixLength int
 
-func parsecidr4(cidr []byte) (uint32, uint32, error) {
-	var mask uint32
-	p := bytes.IndexByte(cidr, '/')
-	if p > 0 {
-		for _, c := range cidr[p+1:] {
-			if c < '0' || c > '9' {
-				return 0, 0, ErrBadIP
-			}
-			mask = mask*10 + uint32(c-'0')
+	// Check for '/' to determine if this is a single IP or prefix. This is the same approach used by net.ParseCIDR
```

Comment:
isn't it just simpler to use `net.ParseCIDR`?

Reply:
I originally was thinking the same thing, but it turns out that the net package is more annoying than that. Using `net.ParseCIDR` works fine, but it is only able to parse prefixes. I thought I could use `net.ParseIP` for addresses, but for some reason it parses all IPv4 addresses as IPv4-mapped IPv6 addresses. I suspect this is why the original author chose to manually parse IPv4 addresses, but used the standard library for IPv6. It also took a while to debug since it fail the `len(address) == net.IPv4Len` check, but debug prints would display an IPv4 value. I can manually perform the check and conversion, but I was unable to see any functions in `net` that could help to do this so it could be a bit messy.

On the other hand, `netip` did have this functionality so it was more straightforward to take that approach. I could use `net.ParseCIDR`, but it seemed cleaner to either use only `net` or `netip` in both cases. I am also guessing `netip` will also be more performant since it appears to only use the stack while parsing unlike `net`, but I have not done any benchmarks to verify this.

Reply:
from playing around with some quick tests: I think `net.ParseCIDR` will not parse a single address, it'll throw an error. More generally: why should we expect a function with this name to handle individual addresses rather than CIDR blocks? If it's to track upstream semantics, best to say.

Reply:
Oh I see. I was basically curious if we could use `net.ParseCIDR` and fall back to whatever you're doing here in case it returns error. But don't have a strong preference as the code seems pretty straightforward.

Reply:
That is correct. The API is built around the idea that it accepts either prefixes or individual IPs as inputs and changing that would require changing most of the existing tests. I'll add a comment to the function explaining that it accepts both.